pkg/reconciler/layer2.go

```diff
@@ -1,61 +1,130 @@
+		item := &layer2List.Items[i]
 		if item.Spec.NodeSelector != nil {
 			selector, err := metav1.LabelSelectorAsSelector(item.Spec.NodeSelector)
 			if err != nil {
 				r.Logger.Error(err, "error converting nodeSelector of layer2 to selector", "layer2", item.ObjectMeta.Name)
-				return nil, err
+				return nil, fmt.Errorf("error converting nodeSelector of layer2 to selector: %w", err)
 			}
 			if !selector.Matches(labels.Set(node.ObjectMeta.Labels)) {
 				r.Logger.Info("local node does not match nodeSelector of layer2", "layer2", item.ObjectMeta.Name, "node", nodeName)
 				continue
 			}
 		}
 
-		l2vnis = append(l2vnis, item)
+		l2vnis = append(l2vnis, *item)
 	}
 
 	return l2vnis, nil
 }
 
 func (r *reconcile) reconcileLayer2(l2vnis []networkv1alpha1.Layer2NetworkConfiguration) error {
-	desired := []nl.Layer2Information{}
+	desired, err := r.createDesired(l2vnis)
+	if err != nil {
+		return err
+	}
+
+	existing, err := r.netlinkManager.ListL2()
+	if err != nil {
+		return fmt.Errorf("error listing L2: %w", err)
+	}
+
+	toDelete := createToDelete(existing, desired)
+
+	create := []nl.Layer2Information{}
+	anycastTrackerInterfaces := []int{}
+	for i := range desired {
+		alreadyExists := false
+		var currentConfig nl.Layer2Information
+		for j := range existing {
+			if desired[i].VlanID == existing[j].VlanID {
+				alreadyExists = true
+				currentConfig = existing[j]
+				break
+			}
+		}
+		if !alreadyExists {
+			create = append(create, desired[i])
+		} else {
+			if err := r.reconcileExistingLayer(&desired[i], &currentConfig, &anycastTrackerInterfaces); err != nil {
+				return err
+			}
+		}
+	}
 
-	for _, layer2 := range l2vnis {
-		spec := layer2.Spec
+	for i := range toDelete {
+		r.Logger.Info("Deleting Layer2 because it is no longer configured", "vlan", toDelete[i].VlanID, "vni", toDelete[i].VNI)
+		errs := r.netlinkManager.CleanupL2(&toDelete[i])
+		for _, err := range errs {
+			r.Logger.Error(err, "Error deleting Layer2", "vlan", toDelete[i].VlanID, "vni", toDelete[i].VNI)
+		}
+	}
+
+	for i := range create {
+		if err := r.createL2(&create[i], &anycastTrackerInterfaces); err != nil {
+			return err
+		}
+	}
+
+	r.anycastTracker.TrackedBridges = anycastTrackerInterfaces
+
+	return nil
+}
+
+func (r *reconcile) createL2(info *nl.Layer2Information, anycastTrackerInterfaces *[]int) error {
+	r.Logger.Info("Creating Layer2", "vlan", info.VlanID, "vni", info.VNI)
+	err := r.netlinkManager.CreateL2(info)
+	if err != nil {
+		return fmt.Errorf("error creating layer2 vlan %d vni %d: %w", info.VlanID, info.VNI, err)
+	}
+	if info.AdvertiseNeighbors {
+		bridgeID, err := r.netlinkManager.GetBridgeID(info)
+		if err != nil {
+			return fmt.Errorf("error getting bridge id for vlanId %d: %w", info.VlanID, err)
+		}
+		*anycastTrackerInterfaces = append(*anycastTrackerInterfaces, bridgeID)
+	}
+	return nil
+}
+
+func (r *reconcile) createDesired(l2vnis []networkv1alpha1.Layer2NetworkConfiguration) ([]nl.Layer2Information, error) {
```

Comment:
maybe call this `getDesired` or something like that. By prefixing everything with `create` it's not clear which functions actually create/apply configuration, and which just create/determine lists of configs.